### PR TITLE
Add in app message lifecycle handler

### PR DESCRIPTION
**Pull request by nan-li**
# Description
## One Line Summary
Adds the In-App Message Lifecycle Handler which includes up to 4 callbacks for the displaying and dismissing of In-App Messages.

### Motivation
This handler was added to give customers visibility into the display lifecycle of IAMs: when an IAM will be displayed, and also when it has been dismissed. 

### Scope
- Extra fields will show up if a user logs the OSInAppMessage instead of just its messageId property, which is fine.
- In-App Messages are only read and not modified.

### Public API Changes
Yes, the documentation will need to be updated for react native to provide these new methods.

### Previous WiP Questions
1. `onWillDismissInAppMessage` is logged before the `inAppMessageClicked` is logged even though clicking on the IAM is what causes it to be dismissed (is this problem?)
2. Can the `InAppMessage` be null? -> no, it should not be
3. The native code does call `sendEvent` for all 4 lifecycle events even if the user has not set all 4. They are just not "caught" by the JavaScript if the user hasn't set the event handler for a particular lifecycle event (native logcat shows all 4).
4. Should the native `OSInAppMessage` classes add `toJSONObject` (Android) or `toDictionary` or `jsonRepresentation` (iOS) as these exist on the other handler/observer objects -> yes, added.
5. Lastly, I noticed that `RNOneSignal` has different implementations for the other handlers and not sure what the differences are for (`NotificationWillShowInForegroundHandler`, `NotificationOpenedHandler`, `ClickHandler`) and if `InAppMessageLifecycleHandler` should be done differently.

# Testing
## Manual testing
Ran the code in example app on Android and iOS emulators.

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [x] In-App Messaging
   - [ ] REST API requests
   - [X] Public API changes

# Checklist
## Overview
   - [X] I have filled out all **REQUIRED** sections above
   - [X] PR does one thing
   - [X] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [X] I have included test coverage for these changes, or explained why they are not needed
   - [X] All automated tests pass, or I explained why that is not possible
   - [X] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [X] Code is as readable as possible.
   - [X] I have reviewed this PR myself, ensuring it meets each checklist item

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/react-native-onesignal/1300)

### CODE REVIEW

**Review by emawby — COMMENTED** [review]
1. Is this the case in the native SDKs as well or does it log in the expected order in native?
2. I don't think the InAppMessage should be null. I can't think of a non-error scenario where it would be null
3. This is fine I think
4. Yes we probably should have the OSInAppMessage toJSONObject in native instead of RN so that we don't need to translate it in every wrapper. This will require native changes and release. Good call!
5. I think your approach here looks good to me, but I will defer to the RN experts.

> *[Reviewable](https://reviewable.io/reviews/onesignal/react-native-onesignal/1300#-:-MmJN7giEP16-0y0qmdA:b-d0kq6x)* status: 0 of 11 files reviewed, all discussions resolved (waiting on @Jeasmine, @jkasten2, and @rgomezp)

**Comment by rgomezp** [comment]
Agree with emawby on all points. Let's make sure to get the native change done with a `toJSONObject()` before getting this out.

File: src/index.d.ts
Comment (rgomezp) [inline src/index.d.ts]: remove large indentation since it there is only a single property

File: src/index.d.ts
Comment (rgomezp) [inline src/index.d.ts]: Nit: the function name ends with "Handler" while the argument is called "handlers". 

For clarity, we should make the naming agree in number.

Since we named the function to match the native method, we should consider modifying the argument. 

One idea is `handlerObject` or `lifecycleObject` or `lifecycleConfig`

Open to other ideas

**Review by rgomezp — COMMENTED** [review]
Looks good other than the small nits and waiting for the changes related to the native JSON stringifier.

File: src/index.d.ts
Comment (nan-li) [reply src/index.d.ts]: Yes, good point. 

I like `handlerObject` since it is an object and having handler in the name matches the method name and the other names we have.

Then should the interface `InAppMessageLifecycleHandlerCallbacks` match the name of the argument too? `InAppMessageLifecycleHandlerObject`?

**Comment by nan-li** [comment]
@emawby @rgomezp I have one more question:

I gave the react-native `InAppMessage` the property of `messageId`. In the native SDKs, they have this property with this same name. However, all the JSON used in the native SDKs pass it around as `"id"` (this exists on `OSInAppMessageInternal`):
```
json[@"id"] = self.messageId
message.messageId = json[@"id"]
````
Should I name the native JSON stringifier for the public `OSInAppMessage` as json[@"messageId"], differing from the `Internal` version, or rename the react-native `InAppMessage` property to `id` instead of `messageId`? 

I believe the JSON keys must match over the bridge?

File: src/index.d.ts
Comment (rgomezp) [reply src/index.d.ts]: Yep that's a good thought, too. Let's do that

**Comment by rgomezp** [comment]
> @emawby @rgomezp I have one more question:
> 
> I gave the react-native `InAppMessage` the property of `messageId`. In the native SDKs, they have this property with this same name. However, all the JSON used in the native SDKs pass it around as `"id"` (this exists on `OSInAppMessageInternal`):
> 
> ```
> json[@"id"] = self.messageId
> message.messageId = json[@"id"]
> ```
> 
> Should I name the native JSON stringifier for the public `OSInAppMessage` as json[@"messageId"], differing from the `Internal` version, or rename the react-native `InAppMessage` property to `id` instead of `messageId`?
> 
> I believe the JSON keys must match over the bridge?

Let's stick to `messageId` since that's what we're doing natively

**Comment by nan-li** [comment]
> Let's stick to `messageId` since that's what we're doing natively

Sounds good. Does this mean the native `OSInAppMessage.toJSONObject()` will have to be `json["messageId"]` instead of `json["id"]`, which differs from `OSInAppMessageInternal.toJSONObject()`?

**Review by jkasten2 — CHANGES_REQUESTED** [review]
> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/onesignal/react-native-onesignal/1300)* status: all files reviewed, 3 unresolved discussions (waiting on @emawby, @Jeasmine, and @nan-li)

___
*[android/src/main/java/com/geektime/rnonesignalandroid/RNUtils.java, line 183 at r1](https://reviewable.io/reviews/onesignal/react-native-onesignal/1300#-Mn2UkPb6hV8zsXKHO-E:-Mn2UkPb6hV8zsXKHO-F:bz8tv86) ([raw file](https://github.com/onesignal/react-native-onesignal/blob/e8c184f91a82bdfd092f9a74923651936d930096/android/src/main/java/com/geektime/rnonesignalandroid/RNUtils.java#L183)):*
> ```Java
>     /**
>      * Converts an OSInAppMessage into a JSON object as OSInAppMessage lacks a toJSONObject() method.
>      * TODO: add toJSONObject() to OSInAppMessage and remove this method.
> ```

Adding a comment here to ensure address this comment before merging.

File: src/index.d.ts
Comment (nan-li) [reply src/index.d.ts]: Changed `InAppMessageLifecycleHandlerCallbacks` to `InAppMessageLifecycleHandlerObject`
Changed `handlers` to `handlerObject`

**Review by jkasten2 — APPROVED** [review]
> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/onesignal/react-native-onesignal/1300)* status: all files reviewed, 1 unresolved discussion (waiting on @emawby, @Jeasmine, @nan-li, and @rgomezp)

**Pull request merged**
